tools/oneclickapps/prepare_digital_ocean_one_click_app_release.py

```diff
@@ -0,0 +1,117 @@
+                incomplete = False
+                break
+        if incomplete:
+            time.sleep(5)
+    droplet.load()
+
+def set_api_request_retry_limits(api_object: digitalocean.baseapi.BaseAPI) -> None:
+    retry = Retry(connect=5, read=5, backoff_factor=0.1)
+    adapter = HTTPAdapter(max_retries=retry)
+    api_object._session.mount('https://', adapter)
+
+def create_droplet(name: str, ssh_keys: List[str], image: str='ubuntu-18-04-x64') -> digitalocean.Droplet:
+    droplet = digitalocean.Droplet(
+        token=manager.token,
+        name=name,
+        region='nyc3',
+        size_slug='s-1vcpu-2gb',
+        image=image,
+        backups=False,
+        ssh_keys=ssh_keys)
+    set_api_request_retry_limits(droplet)
+    droplet.create()
+    sleep_untill_droplet_action_is_completed(droplet, "create")
+    return droplet
+
+def create_ssh_key(name: str, public_key: str) -> digitalocean.SSHKey:
+    action_public_ssh_key_object = digitalocean.SSHKey(name=name, public_key=public_key, token=manager.token)
+    set_api_request_retry_limits(action_public_ssh_key_object)
+    action_public_ssh_key_object.create()
+    return action_public_ssh_key_object
+
+def create_snapshot(droplet: digitalocean.Droplet, snapshot_name: str) -> None:
+    droplet.take_snapshot(snapshot_name, power_off=True)
+    droplet.load()
+    sleep_untill_droplet_action_is_completed(droplet, "snapshot")
+
+def create_dns_records(droplet: digitalocean.Droplet) -> None:
+    domain = digitalocean.Domain(token=manager.token, name='zulipdev.org')
+    set_api_request_retry_limits(domain)
+    domain.load()
+
+    oneclick_test_app_record_names = [TEST_DROPLET_SUBDOMAIN, f"*.{TEST_DROPLET_SUBDOMAIN}"]
+    for record in domain.get_records():
+        if record.name in oneclick_test_app_record_names and record.domain == 'zulipdev.org' and record.type == 'A':
+            record.destroy()
+
+    domain.load()
+    for record_name in oneclick_test_app_record_names:
+        domain.create_new_domain_record(type='A', name=record_name, data=droplet.ip_address)
+
+def setup_one_click_app_installer(droplet: digitalocean.Droplet) -> None:
+    subprocess.call(["fab", "build_image", "-H", droplet.ip_address, "--keepalive", "5"], cwd="marketplace-partners/marketplace_docs/templates/Fabric")
+
+def send_message(content: str) -> None:
+    request = {
+        "type": "stream",
+        "to": os.environ["ZULIP_STREAM"],
+        "topic": "DigitalOcean One Click App",
+        "content": content,
+    }
+    zulip_client.send_message(request)
+
+if __name__ == "__main__":
+    release_version = os.environ["RELEASE_VERSION"]
+
+    generate_ssh_keys()
+    action_public_ssh_key_object = create_ssh_key(f"oneclickapp-{release_version}-image-generator-public-key", get_public_ssh_key())
+
+    image_generator_droplet = create_droplet(f"oneclickapp-{release_version}-image-generator", [action_public_ssh_key_object])
+
+    time.sleep(60)
```

Comment:
This is to wait for ssh to come up, I assume.  Instead of a fixed `sleep`, it seems less flaky to try to make the connection and retry on failure, with a longer (2min?) timeout.

Reply:
Fabric has a `-connection-attempts` flag which defaulted to 1. I set that to 10 instead and removed sleep altogather. Is 10 too much?

https://docs.fabfile.org/en/1.14/usage/fab.html#cmdoption--connection-attempts

There is also a flag to set the timeout seconds which default to 10 seconds. We should probably just keep it as it is.
https://docs.fabfile.org/en/1.14/usage/fab.html#cmdoption--timeout

Reply:
10 means 100 seconds max, which seems fine.  This isn't a particularly time-critical operation, so having it fail slowly is fine.